### DISCUSSION TITLE
The WixUtilExtension ServiceConfig element is not a substitute for the ServiceConfig element

### DISCUSSION THREAD

**Discussion by djohnst10** [wixtoolset/issues#7452] [Questions] [answered]
I converted a WiX 3.11 project to WiX 4.0 and I got the following error when I tried to compile:
```
error WIX1149: ServiceConfig functionality is documented in the Windows Installer SDK to "not [work] as expected." Consider replacing ServiceConfig with the WixUtilExtension ServiceConfig element.
```
I believe this error should not exist because the WixUtilExtension ServiceConfig element is not a replacement for the ServiceConfig element, it is a replacement for the ServiceConfigFailureActions element.

It is true that the [MsiConfigServices Action](https://learn.microsoft.com/en-us/windows/win32/msi/msiconfigureservices-action) does have a warning that states: 

> Windows Services provides the ability to automatically perform predefined actions in response to a failure in a service. To support programmatically setting up these **Recovery Actions** when a service fails, **[MsiServiceConfigFailureActions](https://learn.microsoft.com/en-us/windows/win32/msi/msiserviceconfigfailureactions-table)** was added to MSI in version MSI 5.0. However, this functionality is not working as expected.
>
> To workaround this issue, an application developer should use the **Custom Action** functionality in MSI to run **sc.exe** and set the **Recovery Options** appropriately.

That warning is probably accurate but take note that it is talking about setting up Recovery Actions being enabled by [MsiServiceConfigFailureActions](https://learn.microsoft.com/en-us/windows/win32/msi/msiserviceconfigfailureactions-table), not [MsiServiceConfig](https://learn.microsoft.com/en-us/windows/win32/msi/msiserviceconfig-table) which enables some other stuff like service delay auto start, SIDs, pre shutdown delay, and failure actions which are not enabled by util:ServiceConfig.

From the following snippet which uses ServiceConfig, ServiceConfigFailureActions, and util:ServiceConfig you can see that there are attributes of ServiceConfig that do not have a counterpart in util:ServiceConfig and many attributes in ServiceConfigFailureActions that do have counterparts in util:ServiceConfig.

```xml
<ServiceConfig DelayedAutoStart="yes" OnInstall="yes" OnReinstall="yes" OnUninstall="no" PreShutdownDelay="30000" 
               FailureActionsWhen="failedToStopOrReturnedError" ServiceSid="SERVICE_SID_TYPE_RESTRICTED">
  <RequiredPrivilege>SE_AUDIT_NAME</RequiredPrivilege>
</ServiceConfig>
<ServiceConfigFailureActions OnInstall="yes" OnReinstall="yes" OnUninstall="no"
                             RebootMessage="Service failed. Rebooted!" Command="log.exe 'Service Failure!'"
                             ResetPeriod="86400">
  <Failure Action="restartService" Delay="60" />
  <Failure Action="restartComputer" Delay="120" />
  <Failure Action="runCommand" Delay="180" />
</ServiceConfigFailureActions>
<util:ServiceConfig FirstFailureActionType="restart" SecondFailureActionType="reboot" ThirdFailureActionType="runCommand"
                    RebootMessage="Service failed. Rebooted!" ProgramCommandLine="log.exe 'Service Failure!'"
                    ResetPeriodInDays="1" RestartServiceDelayInSeconds="60" />
```

In short, using the ServiceConfig element should not generate an error. It is possible using the ServiceConfigFailureActions element should generate an error as its functionality seems to be handled by util:ServiceConfig.

**Comment by djohnst10** [comment, marked as answer]
It turned out that I had `<TreatWarningsAsErrors>True</TreatWarningsAsErrors>` in the project file which made the WIX1149 warning an error. I removed TreatWarningsAsErrors and suppressed warning 1149. Now the code compiles (aside from other unrelated errors).

I still think that using ServiceConfig should not be warned against if you are using it for supported and working functionality without an alternative in util:ServiceConfig.

**Reply by robmen** [reply]
I don't think we've interpreted the MSI documentation as narrowly as you have, but maybe your interpretation is correct. If so, we can reduce the scope of the warning. If you know your interpretation of the documentation is correct, opening an issue to change the compiler warning check might be worth it.

**Reply by BMurri** [reply]
Digging through both the documentation as well as searching several reported failures involving those two tables and their associated action, here's what I've found:

If you use the `FailureActionsWhen` attribute, you are [supposed to populate](https://learn.microsoft.com/windows/win32/msi/using-services-configuration) the `ServiceConfigFailureActions` table (the documentation [appears to advise](https://learn.microsoft.com/windows/win32/msi/msiserviceconfig-table) that the rows that attribute writes will have no effect if corresponding entries in the other table aren't present). The `ServiceConfigFailureActions` table is the one implicated in both the reports and the plain reading of the official warning not to use.

So, I would say that the fix is to move the warning in `ParseServiceConfigElement` to any use of the `FailureActionsWhen` attribute and leave the warning in `ParseServiceConfigFailureActionsElement` as it is.

**Reply by MarkStega** [reply]
@robmen Is this in scope for V5?  I know there was discussion about a combined element but I can't find either that discussion nor a V5 features list.

As a total aside I upgraded to V5 RC1 with zero issues but still noted the warning of "\Solutions\OHI\Optimiser.2021\WiX.Services\Product.wxs(370): warning WIX1149: ServiceConfig functionality is documented in the Windows Installer SDK to "not [work] as expected." Consider replacing ServiceConfig with the WixUtilExtension ServiceConfig element."

**Reply by robmen** [reply]
Nope. WiX v5-rc.1 was released last week.
